### PR TITLE
[PORT] Fixes Massive Radio Overtime, Implements a Spatial Grid System for Faster Searching Over Areas

### PR DESCRIPTION
**Pull request by Mat05usz**
# About The Pull Request

Ports: https://github.com/tgstation/tgstation/pull/61422

**If you want to understand how this works, go to the original PR and read it, because sure as hell I won't be able to explain it as nicely as the original author did**

**I ported it due to 2 reasons:** 

1. Faster code good
2. Ike posted it in Projects in stuff to port 

**I'll do a bit of explaining as to why the PR isn't 1:1 with the original yet:**

1. I am aware of things like us having `dview()` and them not ~~(though I'm not sure why is it like that)~~ so I'll be changing several instances of `view()` to `dview()`, unless told otherwise.

2. Furthermore, I'm aware that multiple changes (some not necessarily related to the spatial grid itself, like comment fixes or small loop refactors) haven't been ported by me yet, just need to know if it should be done either in this PR or separate one.

**I also decided to run the benchmark that the original coder provided (or at least part of it), and what I found isn't as optimistic as their testing (in most cases it's still faster than the current solution anyways) BUT that might be because of several reasons:** 

 1. I have no idea what I'm doing, unlike them (most probable)
 2. My PC is shittier (might explain lesser number of iterations, but not the discrepancy in **SOME** of the results)
 3. The test enviroment was much more different (I don't really have info how they benchmarked the code themselves, I just spawned a bunch of humans in different distances from the src, plus some radios around the station and ran the benchmark)

**Screenshots from the benchmarks I did (if they even matter) will be provided soon**

**As for the testing in-game itself:** 

1. I did test the radios on station z-level, on transit z-level etc. Nothing seems to be broken, no runtimes happening, everything is fine (or at least I think so).

_I'm not really sure yet as to what to put in the changelog, if anything_

## Why It's Good For The Game

Faster code good.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

</details>

## Changelog
:cl:
code: Code is now a tiny bit faster
/:cl:

### CODE REVIEW

**Comment by Kapu1178** [comment]
tg has `dview()` but it is unused because HOLY FUCK DVIEW IS SO CURSED

**Comment by Mat05usz** [comment]
> tg has `dview()` but it is unused because HOLY FUCK DVIEW IS SO CURSED

Okay I see it now lmao

Edit: On the other hand, this PR also uses 'mobs' to get over some byond limitations...

**Comment by PowerfulBacon** [comment]
dview() is just view but for when you want to see objects in the dark

**Comment by Mat05usz** [comment]
Reverted https://github.com/BeeStation/BeeStation-Hornet/pull/6587 as the current version of my PR uses `view()` not `dview()` so it takes into account invis mobs, I think.

File: code/__HELPERS/spatial_info.dm [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+				break
+
+/proc/get_hearers_in_radio_ranges(list/obj/item/radio/radios)
+	. = list()
+	// Returns a list of mobs who can hear any of the radios given in @radios
+	for(var/obj/item/radio/radio as anything in radios)
+		. |= get_hearers_in_LOS(radio.canhear_range, radio, FALSE)
+
+///Calculate if two atoms are in sight, returns TRUE or FALSE
+/proc/inLineOfSight(X1,Y1,X2,Y2,Z=1,PX1=16.5,PY1=16.5,PX2=16.5,PY2=16.5)
```
Comment (Archanial) [inline code/__HELPERS/spatial_info.dm]:
```suggestion
/proc/inLineOfSight(X1, Y1, X2, Y2, Z=1, PX1=16.5, PY1=16.5, PX2=16.5, PY2=16.5)
```

File: code/__HELPERS/spatial_info.dm [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+	var/turf/T
+	if(X1==X2)
+		if(Y1==Y2)
+			return TRUE //Light cannot be blocked on same tile
```
Comment (Archanial) [inline code/__HELPERS/spatial_info.dm]:
```suggestion
	if(X1 == X2)
		if(Y1 == Y2)
			return TRUE //Light cannot be blocked on same tile
```

File: code/__HELPERS/spatial_info.dm [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+		if(X1<X2)
+			b+=m
+		while(X1!=X2 || Y1!=Y2)
+			if(round(m*X1+b-Y1))
+				Y1+=signY //Line exits tile vertically
+			else
+				X1+=signX //Line exits tile horizontally
+			T=locate(X1,Y1,Z)
+			if(!T.opacity)
+				return FALSE
```
Comment (Archanial) [inline code/__HELPERS/spatial_info.dm]:
```suggestion
			var/s = SIGN(Y2 - Y1)
			Y1 += s
			while(Y1 != Y2)
				T=locate(X1, Y1, Z)
				if(!T.opacity)
					return FALSE
				Y1 += s
	else
		var/m=(32*(Y2 - Y1)+(PY2 - PY1))/(32*(X2 - X1)+(PX2  -PX1))
		var/b=(Y1 + PY1/32-0.015625)-m*(X1 + PX1/32 - 0.015625) //In tiles
		var/signX = SIGN(X2 - X1)
		var/signY = SIGN(Y2 - Y1)
		if(X1<X2)
			b+=m
		while(X1 != X2 || Y1 != Y2)
			if(round(m*X1 + b - Y1))
				Y1 += signY //Line exits tile vertically
			else
				X1 += signX //Line exits tile horizontally
			T = locate(X1, Y1, Z)
			if(!T.opacity)
				return FALSE
```

File: code/game/atoms_movable.dm [hunk 4]
```diff
@@ -1071,11 +1094,41 @@
 ///allows this movable to hear and adds itself to the important_recursive_contents list of itself and every movable loc its in
 /atom/movable/proc/become_hearing_sensitive(trait_source = TRAIT_GENERIC)
 	if(!HAS_TRAIT(src, TRAIT_HEARING_SENSITIVE))
-		RegisterSignal(src, SIGNAL_REMOVETRAIT(TRAIT_HEARING_SENSITIVE), .proc/on_hearing_sensitive_trait_loss)
+		//RegisterSignal(src, SIGNAL_REMOVETRAIT(TRAIT_HEARING_SENSITIVE), .proc/on_hearing_sensitive_trait_loss)
```
Comment (Archanial) [inline code/game/atoms_movable.dm]: commented code. Actually why is it commented?

**Review by PowerfulBacon — CHANGES_REQUESTED** [review]
Pretty certain this causes new_players to be able to hear in the menu, which allows them to talk and hear noises.
New_players also bypass the chat filter resulting in a lot of N bombs and bans.

File: code/game/atoms_movable.dm (hunk 4 above)
Comment (Mat05usz) [reply code/game/atoms_movable.dm]: Unsure. It's in the original PR.

**Comment by Mat05usz** [comment]
> Pretty certain this causes new_players to be able to hear in the menu, which allows them to talk and hear noises. New_players also bypass the chat filter resulting in a lot of N bombs and bans.

Fixed it by making new_player unable to talk (talked to Bacon about it)

(I realize it seems like an ass fix, tested multiple options and it just makes the most sense. New_player shouldn't be able to talk anyways)

File: code/game/atoms_movable.dm (hunk 4 above)
Comment (Kylerace) [reply code/game/atoms_movable.dm]: we do a little trolling

Comment (Kylerace) [reply code/game/atoms_movable.dm]: was just an oversight as i was working around with important_recursive_contents setters. it can be deleted

**Comment by Kylerace** [comment]
fwiw i did all of my testing in two very specific parts of tramstation which idk if bee has. specifically
Screenshot_1996
Screenshot_1995
when testing get_hearers_in_view() with center as my body. the performance of get_hearers_in_radio_ranges() heavily depends on how radios are distributed across the map, and since you arent porting the unregistering behavior with radios listening to common radio when they cant transmit anyways you arent going to see as good improvement there because youre performing the same work on more radios

File: code/game/atoms_movable.dm (hunk 4 above)
Comment (Mat05usz) [reply code/game/atoms_movable.dm]: Got it, thanks!

**Comment by PowerfulBacon** [comment]
Cannot confirm if it was this or not, but the server had massive TiDi while this was active with get get_hear proc being one of the biggest things.
This might be worse for performance, will confirm by checking TiDi and profiling next round.

**Comment by PowerfulBacon** [comment]
This PR may also be responsible for a headset hard-del in the preference screen

**Review by PowerfulBacon — CHANGES_REQUESTED** [review]
The 2 rounds this was testmerged, time dilation was at like >50% the majority of the round and get_hear was the 5th most expensive proc on the server after hard deletes, parallax, icon2html and lighting.
Also resulted in hard-dels on the preference screen with radio headsets.
Without this PR, the most expensive hearer related proc is can_hear at position #87

Something is definitely wrong.

**Comment by Kylerace** [comment]
@PowerfulBacon any runtimes? get_hear() was just changed from using dview() to 
```
/proc/get_hear(range, atom/source)
	var/lum = source.luminosity
	source.luminosity = 6

	. = view(range, source)
	source.luminosity = lum
```
which wouldnt by itself cause a jump like what youre describing unless maybe it was runtiming a lot

unless youre talking about get_hearers_in_view(), which does use the spatial grid system

**Comment by PowerfulBacon** [comment]
Ah, the jump is most likely from dview() being counted as its own proc so dview was considered less performant rather than get_hear. Checking it again, this seems like nothing changed for get_hear.
Get_hearers_in_view was extremely high however.
Time dilation the 2 rounds it was testmerged was also at >50% the entire time, with the rounds around it being much lower, usually under 20 to 30% with a lot of shit like fires everywhere going on at the same time.
Could have been unrelated, I can try another testmerge at some point.

**Comment by PowerfulBacon** [comment]
Needs more testing

File: code/__DEFINES/maths.dm [hunk 5]
```diff
@@ -28,6 +28,8 @@
 
 #define CEILING(x, y) ( -round(-(x) / (y)) * (y) )
 
+#define ROUND_UP(x) ( -round(-(x)))
```
Comment (PowerfulBacon) [inline code/__DEFINES/maths.dm:31]: Isn't this just ceiling?

Comment (Kylerace) [reply code/__DEFINES/maths.dm:31]: its ceiling with y=1 but simplified

**Comment by Dejaku51** [comment]
rip

**Pull request closed without merging**